### PR TITLE
Fix flutter run cache

### LINKED ISSUE #45269: Flutter run doesn't cache libimobiledevice
## Steps to Reproduce

1. on macOS, delete stamp and cache artifacts for libimobiledevice `rm flutter/bin/cache/libimobiledevice.stamp; rm -rf flutter/bin/cache/artifacts/libimobiledevice`
2. verbosely run a flutter app with only a physical iOS device attached `cd /path/to/flutter/app && flutter run -v`
3. the tool will toolexit because `idevice_id` failed and no devices were found.

## Logs

```
[  +15 ms] executing: /Users/fujino/git/flutter/bin/cache/artifacts/libimobiledevice/idevice_id -h
[   +1 ms] [/Users/fujino/git/flutter/bin/cache/artifacts/libimobiledevice/idevice_id, -h] failed with Invalid argument(s): Cannot find executable for
/Users/fujino/git/flutter/bin/cache/artifacts/libimobiledevice/idevice_id.
[   +7 ms] /usr/bin/xcrun simctl list --json devices
[ +216 ms] No supported devices connected.
[  +10 ms] "flutter run" took 1,437ms.
```

### PR DESCRIPTION
**Pull request by christopherfujino**
## Description

Fixes the issue of `flutter run` not downloading libimobiledevice cache artifacts. The problem was that the tool was checking for attached devices in the verification step, before invoking cache updates, and then exiting with a `ToolExit` when no devices were found (because `idevice_id` was not present). In addition, `requiredArtifacts` for run also depends on libimobiledevice.

The fix the former was to move checking for devices from the verification step (pre-cache check) to the runCommand step (post-cache check) and the fix for the latter was to always update the universal artifacts first, before calling `requiredArtifacts`.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/45269.

## Tests

I added the following tests:

*Replace this with a list of the tests that you added as part of this PR. A change in behaviour with no test covering it
will likely get reverted accidentally sooner or later. PRs must include tests for all changed/updated/fixed behaviors. See [Test Coverage].*

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I signed the [CLA].
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [ ] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter_tools/lib/src/commands/create.dart [hunk 1]
```diff
@@ -337,7 +337,7 @@ class CreateCommand extends FlutterCommand {
         'variable was specified. Unable to find package:flutter.', exitCode: 2);
     }
 
-    await Cache.instance.updateAll(<DevelopmentArtifact>{ DevelopmentArtifact.universal });
+    await Cache.instance.updateAll(<DevelopmentArtifact>{ DevelopmentArtifact.universal }); // TODO is this needed?
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/create.dart]: I don't think so...

File: packages/flutter_tools/lib/src/commands/ide_config.dart [hunk 2]
```diff
@@ -217,7 +217,7 @@ class IdeConfigCommand extends FlutterCommand {
       throwToolExit('Currently, the only supported IDE is IntelliJ\n$usage', exitCode: 2);
     }
 
-    await Cache.instance.updateAll(<DevelopmentArtifact>{ DevelopmentArtifact.universal });
+    await Cache.instance.updateAll(<DevelopmentArtifact>{ DevelopmentArtifact.universal }); // TODO is this needed?
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/ide_config.dart]: I don't think so....

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/commands/ide_config.dart]: Yeah, I tested this, and we can safely remove it.

**Comment by jonahwilliams** [comment]
This should have tests ... you could probably do something like setting up the failing scenario as a pseudo integration test? But with mocks.

**Comment by christopherfujino** [comment]
> This should have tests ... you could probably do something like setting up the failing scenario as a pseudo integration test? But with mocks.

Sounds good. I'm still working on this one, just wanted to get early feedback.

File: packages/flutter_tools/test/general.shard/runner/flutter_command_test.dart [hunk 3]
```diff
@@ -52,7 +52,8 @@ void main() {
     testUsingContext('honors shouldUpdateCache true', () async {
       final DummyFlutterCommand flutterCommand = DummyFlutterCommand(shouldUpdateCache: true);
       await flutterCommand.run();
-      verify(cache.updateAll(any)).called(1);
+      // First call for universal, second for the rest
+      verify(cache.updateAll(any)).called(2);
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/runner/flutter_command_test.dart]: You could verify that each call made sense with `captureAny`

File: packages/flutter_tools/lib/src/runner/flutter_command.dart [hunk 4]
```diff
@@ -589,9 +589,13 @@ abstract class FlutterCommand extends Command<void> {
   Future<FlutterCommandResult> verifyThenRunCommand(String commandPath) async {
     await validateCommand();
 
-    // Populate the cache. We call this before pub get below so that the sky_engine
-    // package is available in the flutter cache for pub to find.
+    // Populate the cache. We call this before pub get below so that the
+    // sky_engine package is available in the flutter cache for pub to find.
     if (shouldUpdateCache) {
+      // First always update universal artifacts, as some of these (e.g.
+      // idevice_id on macOS) are required to run `requiredArtifacts`.
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/runner/flutter_command.dart]: Add a comment on `DevelopmentArtifact.universal` in addition.

**Review by jonahwilliams — APPROVED** [review]
LGTM with two small nits

File: packages/flutter_tools/test/general.shard/runner/flutter_command_test.dart (hunk 3 above)
Comment (christopherfujino) [reply packages/flutter_tools/test/general.shard/runner/flutter_command_test.dart]: cool, didn't know about `captureAny`. Done.

File: packages/flutter_tools/lib/src/runner/flutter_command.dart (hunk 4 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/runner/flutter_command.dart]: Done.

**Pull request merged**